src/utils/UrlUtils.ts

```diff
@@ -0,0 +1,182 @@
+  }
+
+  public static normalizeAsParts(toNormalize: IUrlNormalize): IUrlNormalizedParts {
+    const pathsNormalized = this.normalizePaths(toNormalize);
+    const queryNormalized = this.normalizeQueryString(toNormalize);
+
+    return {
+      pathsNormalized,
+      queryNormalized,
+      path: this.addToUrlIfNotEmpty(pathsNormalized, '/', '')
+    };
+  }
+
+  private static normalizePaths(toNormalize: IUrlNormalize) {
+    return this.toArray(toNormalize.paths).map(path => {
+      if (Utils.isNonEmptyString(path)) {
+        return this.removeProblematicChars(path);
+      }
+      return '';
+    });
+  }
+
+  private static normalizeQueryString(toNormalize: IUrlNormalize) {
+    let queryNormalized: string[] = [];
+
+    if (toNormalize.queryAsString) {
+      const cleanedUp = this.toArray(toNormalize.queryAsString).map(query => {
+        query = this.removeProblematicChars(query);
+        query = this.encodeKeyValuePair(query);
+        return query;
+      });
+      queryNormalized = queryNormalized.concat(cleanedUp);
+    }
+
+    if (toNormalize.query) {
+      const paired: string[][] = pairs(toNormalize.query);
+      const mapped = paired.map(pair => {
+        const [key, value] = pair;
+
+        if (Utils.isNullOrUndefined(value) || Utils.isNullOrUndefined(key)) {
+          return '';
+        }
+
+        if (!this.isEncoded(value)) {
+          return [this.removeProblematicChars(key), Utils.safeEncodeURIComponent(value)].join('=');
+        } else {
+          return [this.removeProblematicChars(key), value].join('=');
+        }
+      });
+      queryNormalized = queryNormalized.concat(mapped);
+    }
+
+    return uniq(queryNormalized);
+  }
+
+  private static addToUrlIfNotEmpty(toAdd: string[], joinWith: string, leadWith: string) {
+    if (Utils.isNonEmptyArray(toAdd)) {
+      return `${leadWith}${compact(toAdd).join(joinWith)}`;
+    }
+    return '';
+  }
+
+  private static startsWith(searchString: string, targetString: string) {
+    return targetString.charAt(0) == searchString;
+  }
+
+  private static endsWith(searchString: string, targetString: string) {
+    return targetString.charAt(targetString.length - 1) == searchString;
+  }
+
+  private static removeAtEnd(searchString: string, targetString: string) {
+    while (this.endsWith(searchString, targetString)) {
+      targetString = targetString.slice(0, targetString.length - searchString.length);
+    }
+
+    return targetString;
+  }
+
+  private static removeAtStart(searchString: string, targetString: string) {
+    while (this.startsWith(searchString, targetString)) {
+      targetString = targetString.slice(searchString.length);
+    }
+    return targetString;
+  }
+
+  private static toArray(parameter: string | string[]): string[] {
+    let ret: string[];
+    if (!isArray(parameter)) {
+      ret = [parameter];
+    } else {
+      ret = parameter;
+    }
+    return ret;
+  }
+
+  private static encodeKeyValuePair(pair: string) {
```

Comment:
In cases where `split.length == 0` or `key` or `value` are not defined, we return `pair`. I'm not sure but it seems like we would be failing silently in such cases. What about returning an empty string?